### PR TITLE
feat: Feature/melodic merge master 15 07 2021

### PR DESCRIPTION
**Pull request by 27Apoorva**

### CODE REVIEW

File: amcl/src/amcl/pf/pf.c [hunk 1]
```diff
@@ -144,34 +142,23 @@ void pf_init(pf_t *pf, int count, pf_vector_t* means, pf_matrix_t* covs)
...
-	  }
-
-	  // Compute the new sample poses
-	  for (i = 0; i < samples; i++)
-	  {
-		sample = set->samples + sampleIndex++;
-		sample->weight = 1.0 / pf->max_samples;
-		sample->pose = pf_pdf_gaussian_sample(pdf);
+    sample = set->samples + i++;
+    sample->weight = 1.0 / pf->max_samples;
```
Comment (DrCodeWizzard) [inline amcl/src/amcl/pf/pf.c:152]: possible divide by zero

File: amcl/src/amcl_node.cpp [hunk 2]
```diff
@@ -277,7 +274,7 @@ class AmclNode
 
     ros::Publisher brainstem_driver_pose_pub_;
 
-    std::vector<amcl_hyp_t> initial_poses_hyp_;
+    amcl_hyp_t* initial_pose_hyp_;
```
Comment (DrCodeWizzard) [inline amcl/src/amcl_node.cpp:277]: you should avoid using raw pointers... the vector is a better choice...

File: amcl/src/amcl_node.cpp [hunk 3]
```diff
@@ -365,7 +362,7 @@ AmclNode::AmclNode() :
         odom_(NULL),
         laser_(NULL),
 	      private_nh_("~"),
-        initial_poses_hyp_(),
+        initial_pose_hyp_(NULL),
```
Comment (DrCodeWizzard) [inline amcl/src/amcl_node.cpp:365]: use nullptr instead of NULL

File: amcl/src/amcl_node.cpp [hunk 4]
```diff
@@ -1739,20 +1705,12 @@ bool
...
-	}
-
-    pf_init(pf_, initial_poses_hyp_.size(), &initial_means[0], &initial_covs[0]);
+  if (initial_pose_hyp_ !=NULL && map_ != NULL ) {
+    pf_init(pf_, initial_pose_hyp_->pf_pose_mean, initial_pose_hyp_->pf_pose_cov);
     pf_init_ = false;
 
-    initial_poses_hyp_.erase(initial_poses_hyp_.begin(), initial_poses_hyp_.end());
+    delete initial_pose_hyp_;
+    initial_pose_hyp_ = NULL;
```
Comment (DrCodeWizzard) [inline amcl/src/amcl_node.cpp:1713]: memory leak!!!

**Review by DrCodeWizzard — CHANGES_REQUESTED** [review]

**Comment by 27Apoorva** [comment]
comments will be addressed by https://6river.atlassian.net/browse/MAP-231 in next release

**Review by DrCodeWizzard — APPROVED** [review]

**Pull request merged**